### PR TITLE
[exporter/datadog] 🧰 Honor configuration hostname when there are multiple exporters

### PR DESCRIPTION
**Pull request by mx-psi**
**Description:** 

Fixes a bug that happens when two Datadog exporters with different hostnames are defined, e.g. for shipping to two different sites:

```yaml
exporter:
  datadog/us:
    hostname: host-us
    api:
      key: api-key-us

  datadog/eu:
    hostname: host-eu
    api:
      key: api-key-eu
      site: datadoghq.eu
```

In this situation, with the previous code, either `host-us` or `host-eu` would eventually be applied to both exporters (whichever ended up in the cache last). We now store only the system/EC2 hostnames in the cache, avoiding this issue.

The PR also checks for hostname validity during configuration sanitization, since we would otherwise spam the logs with error messages.

**Link to tracking Issue:** n/a

**Testing:** Added unit tests

**Documentation:** n/a

### CODE REVIEW

**Review by albertvaka — APPROVED** [review]
👍 from Datadog.

**Comment by mx-psi** [comment]
@open-telemetry/collector-contrib-maintainer PTAL, this has a review from Datadog so I assume it can be merged if it looks fine without an approver review

**Review by tigrannajaryan — APPROVED** [review]

**Pull request merged**
